### PR TITLE
fix(agent): prevent JSON serialization errors with non-serializable direct tool parameters

### LINKED ISSUE #350: [BUG] Agent object not JSON serializable when passed to direct tool calls
### Checks

- [x] I have updated to the lastest minor and patch version of Strands
- [x] I have checked the documentation and this is not expected behavior
- [x] I have searched [./issues](./issues?q=) and there are no duplicates of my issue

### Strands Version

source

### Python Version

source

### Operating System

macOS

### Installation Method

git clone

### Steps to Reproduce

When passing an `Agent` object as a parameter to direct tool calls (e.g., `agent.tool.use_llm(agent=agent)`), the SDK crashes with a `TypeError: Object of type Agent is not JSON serializable` error.

## Problem Description
The issue occurs in the `_record_tool_execution` method in `agent.py` at line 547 when trying to serialize tool input parameters using `json.dumps(tool['input'])`. If any of the tool parameters contains an `Agent` object, the JSON serialization fails.

## Error Details
```
TypeError: Object of type Agent is not JSON serializable
```

**Stack trace location:**
```python
File "/Users/cagatay/tools/.venv/lib/python3.13/site-packages/strands/agent/agent.py", line 547, in _record_tool_execution
    {"text": (f"agent.tool.{tool['name']} direct tool call.\nInput parameters: {json.dumps(tool['input'])}\n")}
```

## Reproduction Steps
1. Create an agent with tools that accept an `agent` parameter (like `use_llm`)
2. Call the tool directly passing the agent object: `agent.tool.use_llm(agent=agent, ...)`
3. The SDK attempts to record the tool execution and fails when serializing the input parameters

## Minimal Reproduction Code
```python
from strands import Agent
from strands_tools import use_llm

agent = Agent(tools=[use_llm])

# This will fail with JSON serialization error
result = agent.tool.use_llm(
    prompt="Test prompt",
    system_prompt="You are a helper.",
    tools=["calculator"],
    agent=agent  # This Agent object causes the JSON serialization to fail
)
```

### Expected Behavior

Direct tool calls should work without crashing, even when passing complex objects like Agent instances as parameters.

### Actual Behavior

```
result = agent.tool.use_llm(
    prompt="Calculate 2 + 2 and read a file",
    system_prompt="You are a helper.",
    tools=["calculator", "file_read"],  # Only these 2 tools will be available
    agent=agent
)

print(result)
# (.venv) c² ~/tools/ [main] python3 test.py
# I can help you with both tasks. Let me break them down:

# 1. Calculating 2 + 2 is straightforward, I can use the calculator tool for this.
# 2. For reading a file, I'll need more information about which file you want me to read and how you want me to read it.

# Let me first calculate 2 + 2:
# Tool #1: calculator
# ╭──────────────────────────────────────── Calculation Result ─────────────────────────────────────────╮
# │                                                                                                     │
# │  ╭───────────┬─────────────────────╮                                                                │
# │  │ Operation │ Evaluate Expression │                                                                │
# │  │ Input     │ 2 + 2               │                                                                │
# │  │ Result    │ 4                   │                                                                │
# │  ╰───────────┴─────────────────────╯                                                                │
# │                                                                                                     │
# ╰─────────────────────────────────────────────────────────────────────────────────────────────────────╯
# For reading a file, I need more information:
# 1. Which file would you like me to read? Please provide a path.
# 2. How would you like to read it? Options include:
#    - view: Display the entire file content
#    - lines: Show specific line ranges
#    - preview: Quick content preview
#    - search: Find specific text in the file
#    - and other modes

# Please let me know the file path and preferred reading mode, and I'll be happy to help.Traceback (most recent call last):
#   File "/Users/cagatay/tools/test.py", line 6, in <module>
#     result = agent.tool.use_llm(
#         prompt="Calculate 2 + 2 and read a file",
#     ...<2 lines>...
#         agent=agent
#     )
#   File "/Users/cagatay/tools/.venv/lib/python3.13/site-packages/strands/agent/agent.py", line 147, in caller
#     self._agent._record_tool_execution(
#     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
#         tool_use, tool_result, user_message_override, self._agent.messages
#         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
#     )
#     ^
#   File "/Users/cagatay/tools/.venv/lib/python3.13/site-packages/strands/agent/agent.py", line 547, in _record_tool_execution
#     {"text": (f"agent.tool.{tool['name']} direct tool call.\nInput parameters: {json.dumps(tool['input'])}\n")}
#                                                                                 ~~~~~~~~~~^^^^^^^^^^^^^^^
#   File "/opt/homebrew/Cellar/python@3.13/3.13.5/Frameworks/Python.framework/Versions/3.13/lib/python3.13/json/__init__.py", line 231, in dumps
#     return _default_encoder.encode(obj)
#            ~~~~~~~~~~~~~~~~~~~~~~~^^^^^
#   File "/opt/homebrew/Cellar/python@3.13/3.13.5/Frameworks/Python.framework/Versions/3.13/lib/python3.13/json/encoder.py", line 200, in encode
#     chunks = self.iterencode(o, _one_shot=True)
#   File "/opt/homebrew/Cellar/python@3.13/3.13.5/Frameworks/Python.framework/Versions/3.13/lib/python3.13/json/encoder.py", line 261, in iterencode
#     return _iterencode(o, 0)
#   File "/opt/homebrew/Cellar/python@3.13/3.13.5/Frameworks/Python.framework/Versions/3.13/lib/python3.13/json/encoder.py", line 180, in default
#     raise TypeError(f'Object of type {o.__class__.__name__} '
#                     f'is not JSON serializable')
# TypeError: Object of type Agent is not JSON serializable
# (.venv) c² ~/tools/ [main] 
```

### Additional Context

This bug affects any tool that accepts complex objects as parameters, not just the `use_llm` tool. The issue is in the core agent recording mechanism rather than specific tools.

### Option 1: Filter non-serializable objects
Modify the `_record_tool_execution` method to filter out non-JSON-serializable objects before calling `json.dumps()`:

```python
def _record_tool_execution(self, tool: ToolUse, tool_result: ToolResult, user_message_override: Optional[str], messages: Messages) -> None:
    # Filter out non-serializable objects
    serializable_input = {}
    for key, value in tool['input'].items():
        try:
            json.dumps(value)  # Test if serializable
            serializable_input[key] = value
        except (TypeError, ValueError):
            serializable_input[key] = f"<non-serializable: {type(value).__name__}>"
    
    user_msg_content: List[ContentBlock] = [
        {"text": (f"agent.tool.{tool['name']} direct tool call.\nInput parameters: {json.dumps(serializable_input)}\n")}
    ]
    # ... rest of the method
```

### Option 2: Use a custom JSON encoder
Create a custom JSON encoder that handles Agent objects and other non-serializable types gracefully.

### Option 3: Skip parameter logging for complex objects
Add a flag or configuration to skip detailed parameter logging when complex objects are involved.

### Related Issues

_No response_

### PR DESCRIPTION
**Pull request by cagataycali**
## Description

This PR fixes a critical issue where passing non-serializable objects as tool parameters would cause JSON serialization errors during tool call recording, leading to crashes when agents tried to record their tool usage.

### The Problem
When agents passed complex objects like Agent instances, custom classes, functions, or other non-JSON-serializable objects as tool parameters, the `_record_tool_execution()` method would fail with cryptic JSON serialization errors, breaking the entire tool execution flow.

### The Solution
- **Smart Parameter Filtering**: Added logic to test each parameter with `json.dumps()` before recording
- **Graceful Degradation**: Replace non-serializable objects with descriptive strings like `<non-serializable: TypeName>`
- **Maintain Functionality**: Tool execution continues normally while providing clear indicators in the recorded history
- **Clean Implementation**: No breaking changes to existing functionality

### Key Changes
- Add parameter filtering in `_record_tool_execution()` to handle non-serializable objects
- Remove unused `messages` parameter from method signature
- Fix message format consistency (`agent.tool.{name}` vs `agent.{name}`)
- Comprehensive test coverage for edge cases and regression prevention

## Related Issues

Closes #350

## Documentation PR

No documentation changes needed - this is an internal bug fix that maintains existing API.

## Type of Change

Bug fix

## Testing

Comprehensive test suite added covering:

- [x] Single non-serializable objects (Agent instances)
- [x] Multiple types of non-serializable objects (custom classes, functions, sets, complex numbers)
- [x] Edge cases with nested structures and None values
- [x] Normal serializable parameters (regression testing)
- [x] Disabled recording scenarios
- [x] I ran `hatch run prepare` - all pre-commit hooks passed

### Test Results
All existing tests pass, and 5 new test functions added:
- `test_agent_tool_non_serializable_parameter_filtering`
- `test_agent_tool_multiple_non_serializable_types`
- `test_agent_tool_serialization_edge_cases`
- `test_agent_tool_no_non_serializable_parameters`
- `test_agent_tool_record_direct_tool_call_disabled_with_non_serializable`

## Checklist
- [x] I have read the CONTRIBUTING document
- [x] I have added any necessary tests that prove my fix is effective or my feature works
- [x] I have updated the documentation accordingly (none needed for internal bug fix)
- [x] I have added an appropriate example to the documentation to outline the feature, or no new docs are needed
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published (none required)

----

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by yakshpradippatel — APPROVED** [review]

File: src/strands/agent/agent.py [hunk 1]
```diff
@@ -617,11 +614,24 @@ def _record_tool_execution(
             tool: The tool call information.
             tool_result: The result returned by the tool.
             user_message_override: Optional custom message to include.
-            messages: The message history to append to.
         """
         # Create user message describing the tool call
+
+        # Filter out non-serializable objects to prevent JSON serialization errors
```
Comment (mkmeral) [inline src/strands/agent/agent.py]: I don't like this approach of serializing everything one by one. Instead we can use `default`. See https://stackoverflow.com/a/56138540

Comment (cagataycali) [reply src/strands/agent/agent.py]: That's a good callout

File: src/strands/agent/agent.py [hunk 2]
```diff
@@ -617,11 +614,24 @@ def _record_tool_execution(
...
+        serializable_input = {}
+        for key, value in tool["input"].items():
+            try:
+                json.dumps(value)  # Test if serializable
+                serializable_input[key] = value
+            except (TypeError, ValueError):
+                serializable_input[key] = f"<non-serializable: {type(value).__name__}>"
```
Comment (JackYPCOnline) [inline src/strands/agent/agent.py]: let's suppose a tool["input"].items() has two non- serializable items have same key with different value.

Comment (cagataycali) [reply src/strands/agent/agent.py]: Thanks, with new changes we're relying on default handling of json.dumps ^^
